storage/statedb/hasher.go

```diff
@@ -76,226 +75,126 @@ func returnHasherToPool(h *hasher) {
-
-	case *fullNode:
-		// Hash the full node's children, caching the newly hashed subtrees
-		collapsed, cached := n.copy(), n.copy()
-
-		for i := 0; i < 16; i++ {
-			if n.Children[i] != nil {
-				collapsed.Children[i], cached.Children[i] = h.hash(n.Children[i], db, false)
-			}
-		}
-		cached.Children[16] = n.Children[16]
-		return collapsed, cached
-
+		collapsed, cached := h.hashFullNodeChildren(n)
+		hashed = h.fullnodeToHash(collapsed, force)
+		if hn, ok := hashed.(hashNode); ok {
+			cached.flags.hash = hn
+		} else {
+			cached.flags.hash = nil
+		}
+		return hashed, cached
 	default:
 		// Value and hash nodes don't have children so they're left as were
-		return n, original
+		return n, n
 	}
 }
 
-type hashResult struct {
-	index     int
-	collapsed node
-	cached    node
+// hashShortNodeChildren collapses the short node. The returned collapsed node
+// holds a live reference to the Key, and must not be modified.
+// The cached
+func (h *hasher) hashShortNodeChildren(n *shortNode) (collapsed, cached *shortNode) {
+	// Hash the short node's child, caching the newly hashed subtree
+	collapsed, cached = n.copy(), n.copy()
+	// Previously, we did copy this one. We don't seem to need to actually
+	// do that, since we don't overwrite/reuse keys
+	// cached.Key = common.CopyBytes(n.Key)
+	collapsed.Key = hexToCompact(n.Key)
+	// Unless the child is a valuenode or hashnode, hash it
+	switch n.Val.(type) {
+	case *fullNode, *shortNode:
+		collapsed.Val, cached.Val = h.hash(n.Val, false)
+	}
+	return collapsed, cached
 }
 
-func (h *hasher) hashChildrenFromRoot(original node, db *Database) (node, node) {
-	switch n := original.(type) {
-	case *shortNode:
-		// Hash the short node's child, caching the newly hashed subtree
-		collapsed, cached := n.copy(), n.copy()
-		collapsed.Key = hexToCompact(n.Key)
-		cached.Key = common.CopyBytes(n.Key)
-
-		if _, ok := n.Val.(valueNode); !ok {
-			collapsed.Val, cached.Val = h.hash(n.Val, db, false)
+func (h *hasher) hashFullNodeChildren(n *fullNode) (collapsed *fullNode, cached *fullNode) {
+	// Hash the full node's children, caching the newly hashed subtrees
+	cached = n.copy()
+	collapsed = n.copy()
+	for i := 0; i < 16; i++ {
+		if child := n.Children[i]; child != nil {
+			collapsed.Children[i], cached.Children[i] = h.hash(child, false)
+		} else {
+			collapsed.Children[i] = nilValueNode
 		}
-		return collapsed, cached
-
-	case *fullNode:
-		// Hash the full node's children, caching the newly hashed subtrees
-		collapsed, cached := n.copy(), n.copy()
-
-		hashResultCh := make(chan hashResult, 16)
-		numRootChildren := 0
-		for i := 0; i < 16; i++ {
-			if n.Children[i] != nil {
-				numRootChildren++
-				go func(i int, n node) {
-					childHasher := newHasher(h.onleaf)
-					defer returnHasherToPool(childHasher)
-					collapsedFromChild, cachedFromChild := childHasher.hash(n, db, false)
-					hashResultCh <- hashResult{i, collapsedFromChild, cachedFromChild}
-				}(i, n.Children[i])
-			}
-		}
```

Comment:
This part is optimized in order to calculate root hash more efficiently. Was this considered?

The related PR: https://github.com/klaytn/klaytn/pull/310
cc @kjhman21 @ehnuje